File: src/main/kotlin/situatedInference/ContextResolver.java
package situatedInference;

import com.ontotext.trree.AbstractRepositoryConnection;
import com.ontotext.trree.StatementIdIterator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import proof.ProofPlugin;

import java.util.HashSet;
import java.util.Set;

public class ContextResolver {

    private final AbstractRepositoryConnection repositoryConnection;
    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    public ContextResolver(AbstractRepositoryConnection repositoryConnection) {
        this.repositoryConnection = repositoryConnection;
    }

    public Set<Quad> getStatementWithAllContexts(Quad statement) {
//        Logger logger = LoggerFactory.getLogger(this.getClass());
        Set<Quad> statementWithAllContexts = new HashSet<Quad>();
        statementWithAllContexts.add(new Quad(statement.subject, statement.predicate, statement.object, statement.context, statement.status));
        try (StatementIdIterator ctxIter = repositoryConnection.getStatements(statement.subject, statement.predicate, statement.object, true, 0, ProofPlugin.excludeDeletedHiddenInferred)) {
            logger.debug(String.format("Contexts for %d %d %d", statement.subject, statement.predicate, statement.object));
            while (ctxIter.hasNext()) {
                statementWithAllContexts.add(new Quad(ctxIter.subj, ctxIter.pred, ctxIter.obj, ctxIter.context, ctxIter.status));
                logger.debug(String.valueOf(ctxIter.context));
                ctxIter.next();
            }
            logger.debug("All contexts for antecedent" + statementWithAllContexts);
        }
        return statementWithAllContexts;
    }
}

File: src/main/kotlin/situatedInference/ContextResolver.kt
package situatedInference

import com.ontotext.trree.AbstractRepositoryConnection
import com.ontotext.trree.StatementIdIterator.*
import com.ontotext.trree.sdk.Entities.UNBOUND
import org.slf4j.LoggerFactory

const val excludeDeletedHiddenInferred =
    DELETED_STATEMENT_STATUS or SKIP_ON_BROWSE_STATEMENT_STATUS or INFERRED_STATEMENT_STATUS

class ContextResolver(private val repositoryConnection: AbstractRepositoryConnection) {
    private val logger = LoggerFactory.getLogger(this.javaClass)

//    fun getStatementWithAllContexts(statement: Quad): Set<Quad> {
//        val statementWithAllContexts = mutableSetOf(statement)
//        repositoryConnection.getStatements(
//            statement.subject,
//            statement.predicate,
//            statement.`object`,
//            UNBOUND,
//            excludeDeletedHiddenInferred
//        ).use { ctxIter ->
//            logger.debug("Contexts for {} {} {}", statement.subject, statement.predicate, statement.`object`)
//
//            while (ctxIter.hasNext()) {
//                statementWithAllContexts.add(Quad(ctxIter))
//                logger.debug(ctxIter.context.toString())
//                ctxIter.next()
//            }
//
//            logger.debug("All contexts for antecedent {}", statementWithAllContexts)
//        }
//        return statementWithAllContexts
//    }

    fun getStatementWithAllContexts(statement: Quad): Set<Quad> {
        return repositoryConnection.getStatements(
            statement.subject,
            statement.predicate,
            statement.`object`,
            UNBOUND,
            excludeDeletedHiddenInferred
        ).asSequence().toSet().also {
            logger.debug("Contexts for {} {} {} = {}", statement.subject, statement.predicate, statement.`object`, it)
        }
    }
}